vpc/service/assign_addresses_v3.go

```diff
@@ -1115,34 +1116,138 @@ func getBorderGroupForENI(ctx context.Context, tx *sql.Tx, eni *ec2.NetworkInter
+		span.SetStatus(traceStatusFromError(err))
+		return nil, err
+	}
+
+	logger.G(ctx).WithField("usedIPAddresses", usedIPAddresses.List()).WithField("allInterfaceIPAddresses", allInterfaceIPAddresses.List()).Debug("Trying to assign IPv6 Address")
+	unusedIPAddresses := allInterfaceIPAddresses.Difference(usedIPAddresses)
+
+	if l := unusedIPAddresses.Len(); l > 0 {
+		unusedIPv6AddressesList := unusedIPAddresses.List()
+
+		row := tx.QueryRowContext(ctx, "SELECT ip_address FROM ip_last_used_v3 WHERE ip_address = any($1::inet[]) AND vpc_id = $2 ORDER BY last_seen ASC LIMIT 1", pq.Array(unusedIPv6AddressesList), aws.StringValue(branchENI.VpcId))
+		var ipAddress string
+		err = row.Scan(&ipAddress)
+		if err == sql.ErrNoRows {
+			// Effectively choose a random one.
+			ipAddress = unusedIPv6AddressesList[rand.Intn(l)] // nolint: gosec
+		} else if err != nil {
+			err = errors.Wrap(err, "Could not fetch utilized IPv6 addresses from the database")
+			tracehelpers.SetStatus(err, span)
+			return nil, err
+		}
+
+		return &vpcapi.UsableAddress{
+			Address: &vpcapi.Address{
+				Address: ipAddress,
+			},
+			PrefixLength: uint32(128),
+		}, nil
+	}
+
+	ipsToAssign := maxIPAddresses - allInterfaceIPAddresses.Len()
+	if ipsToAssign <= 0 {
+		err = status.Errorf(codes.FailedPrecondition, "%d IPv4 addresses already assigned to interface", allInterfaceIPAddresses.Len())
+		span.SetStatus(traceStatusFromError(err))
+		return nil, err
+	}
+	if ipsToAssign > batchSize {
+		ipsToAssign = batchSize
+	}
+
+	assignIpv6AddressesInput := ec2.AssignIpv6AddressesInput{
+		NetworkInterfaceId: branchENI.NetworkInterfaceId,
+	}
+
+	row := tx.QueryRowContext(ctx, "SELECT cidr6 FROM subnets WHERE subnet_id = $1", aws.StringValue(branchENI.SubnetId))
+	var subnetCIDR sql.NullString
+	err = row.Scan(&subnetCIDR)
+	if err != nil {
+		err = fmt.Errorf("Cannot scan cidr6 from subnet %s: %w", aws.StringValue(branchENI.SubnetId), err)
 		tracehelpers.SetStatus(err, span)
 		return nil, err
 	}
 
-	_, net, err := net.ParseCIDR(prefix)
+	if !subnetCIDR.Valid {
+		err = fmt.Errorf("Subnet %s does not have cidr6", aws.StringValue(branchENI.SubnetId))
+		tracehelpers.SetStatus(err, span)
+		return nil, err
+	}
+
+	_, ipnet, err := net.ParseCIDR(subnetCIDR.String)
 	if err != nil {
-		err = fmt.Errorf("Cannot parse prefix %q: %w", prefix, err)
+		err = fmt.Errorf("Cannot parse prefix %q: %w", subnetCIDR.String, err)
 		tracehelpers.SetStatus(err, span)
 		return nil, err
 	}
 
-	prefixInt := cidr.IPv6tod(net.IP)
-	newIPInt := prefixInt.Add(prefixInt, big.NewInt(lastAssigned))
-	newIP := cidr.DtoIPv6(newIPInt)
+	var counter uint32
+	r := rand.New(rand.NewSource(time.Now().UnixNano())) // nolint: gosec
+	atomic.StoreUint32(&counter, r.Uint32())
+	actual, _ := vpcService.counters.LoadOrStore(aws.StringValue(branchENI.SubnetId), &counter)
+	counterPointer := actual.(*uint32)
+
+	base := atomic.AddUint32(counterPointer, uint32(ipsToAssign)) - uint32(ipsToAssign)
```

Comment:
I think this was mentioned earlier but whats the reason to abandon the earlier method?

Reply:
Which earlier method? We cannot rely on using random assign IP, due to our scale and limitations with AWS.